src/deepsparse/benchmark/torchscript_engine.py

```diff
@@ -0,0 +1,325 @@
+
+import logging
+import time
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy
+
+
+try:
+    import torch
+
+    torch_import_error = None
+except Exception as torch_import_err:
+    torch_import_error = torch_import_err
+    torch = None
+
+try:
+    import onnxruntime
+
+    ort_import_error = None
+except Exception as ort_import_err:
+    onnxruntime = None
+    ort_import_error = ort_import_err
+
+try:
+    # flake8: noqa
+    from deepsparse.cpu import cpu_architecture
+except ImportError:
+    raise ImportError(
+        "Unable to import deepsparse python apis. "
+        "Please contact support@neuralmagic.com"
+    )
+
+__all__ = ["TorchScriptEngine"]
+
+_LOGGER = logging.getLogger(__name__)
+
+ARCH = cpu_architecture()
+NUM_CORES = ARCH.num_available_physical_cores
+
+
+def _validate_torch_import():
+    if ort_import_error is not None:
+        raise ImportError(
+            "An exception occurred when importing onxxruntime. Please verify that "
+            "onnxruntime is installed in order to use the onnxruntime inference "
+            "engine.\n\n`onnxruntime` can be installed by running the command "
+            "`pip install deepsparse[onnxruntime]`"
+            f"\n\nException info: {ort_import_error}"
+        )
+
+
+def _validate_batch_size(batch_size: int) -> int:
+    if batch_size < 1:
+        raise ValueError("batch_size must be greater than 0")
+
+    return batch_size
+
+
+def _select_device(device: str):
+    device = str(device).lower()
+    if device == "cuda" and torch.cuda.is_available():
+        return "cuda"
+    return "cpu"
+
+
+def _validate_jit_model(model):
+    if isinstance(model, torch.jit.ScriptModule):
+        return
+    raise ValueError(f"{model} is not a torch.jit model")
+
+
+class TorchScriptEngine(object):
+    """
+    Given a loaded Torchscript(.pt) model or its saved file path, create an
+     that compiles the given pytorch file,
+
+    # Note 1: Engines are compiled for a specific batch size
+
+    # :param model: Either a path to the model's .pt file or the loaded model
+    # :param batch_size: The batch size of the inputs to be used with the engine
+    # :param device: Hardware to run the engine on. Either cpu or cuda
+    """
+
+    def __init__(
+        self,
+        model: Union[str, "Model"],
+        batch_size: int = 1,
+        device: str = "cpu",
+        **kwargs,
+    ):
+        if torch is None:
+            raise ImportError(f"Unable to import torch, error: {torch_import_error}")
+
+        _validate_torch_import()
+
+        self._batch_size = _validate_batch_size(batch_size)
+        self._device = _select_device(device)
+
+        if isinstance(model, torch.nn.Module):
+            self._model = torch.jit.script(model).eval()
+        elif isinstance(model, str):
+            self._model = torch.jit.load(model).eval()
+        else:
+            self._model = model
+
+        _validate_jit_model(self._model)
```

Comment:
IMO we don't need to convert `Module` objects into `jit` modules, it's fine to keep them as they are, need to update the validation accordingly (ie both `Module` and `ScriptModule` valid)